### DISCUSSION TITLE
Setup new card from scratch

### DISCUSSION THREAD

**Discussion by tjboldt** [tjboldt/Apple2-IO-RPi#63] [Announcements]
The instructions have been moved to the code...

For the classic card I designed, the setup is here:
https://github.com/tjboldt/Apple2-IO-RPi/blob/main/ClassicSetup.md

For the A2Pico card designed by Ralle and Oliver, the setup is here:
https://github.com/tjboldt/Apple2-IO-RPi/blob/main/PicoSetup.md

**Comment by tjboldt** [comment]
Note: Updated instructions to use Pi Imager advanced settings to specify ssh and wifi when creating the microSD card image and to switch to a systemd service instead of creating a cronjob.

**Reply by tjboldt** [reply]
Replaced steps 13 through 29 with a single step to just run a script to install everything

**Comment by JPT77** [comment]
Hi
I made everything except step "11. Install firmware with utility."
so, the device does nothing without that?
how do I install the firmware when I cannot access the disk image from the apple2?
I found the "utility" sources in RaspberryPi/driveimage/ but not the hdv image.
could you please provide a floppy image for transfer with adtpro?

thank you very much

**Reply by tjboldt** [reply]
I usually write the EEPROM directly with a programmer but it is possible to write it with the device itself. The drive image used to be a file but recently it's be dynamically generated. You can use the ProDOS-Utilities to create a drive with the latest code or you can grab the old binary image (which should have the same firmware and update utility as the current version). https://github.com/tjboldt/Apple2-IO-RPi/blob/ed2acd9d5fe7c313c47504837055f4a613ca2088/RaspberryPi/Apple2-IO-RPi.hdv

**Reply by microbmen** [reply]
[applepi.zip](https://github.com/tjboldt/Apple2-IO-RPi/files/11067919/applepi.zip)
@JPT77 I made this disk which has the firmware bin and and basic code saved as 'Update'.  Just boot the disk and run update.

**Reply by JPT77** [reply]
thanks @microbmen 
writing from the apple failed. well it told it succeeded. but it wasn't written correctly. I am suspecting an issue with my apple...

**Reply by k6lcm** [reply]
I too am not able to startup from the device that I built from scratch. Since I don't have the capability to write an EEPROM directly, what are the steps to do this from the device itself? When I power on my Apple IIe with the card installed, it just starts up to a BASIC prompt and doesn't seem to recognize the device.

**Reply by JPT77** [reply]
Write the above applepi.zip to a floppy disk, boot it and run the update. 
If it does not boot, try PR#slotnumber
My Apple II+ is not able to autmatically boot from the device, I always have to use PR#5.

**Reply by k6lcm** [reply]
Hmmm. I tried booting the applepi.do image using my FloppyEMU. For me on my IIe, that disk image just boots into a ProDOS BASIC 1.6 screen with an empty BASIC prompt.

**Reply by k6lcm** [reply]
Doh! I typed RUN UPDATE and it worked! I'll let you know how it goes.

**Reply by JPT77** [reply]
@tjboldt could your please add the update.zip to your instructions above? make step 11 a little bit more elaborate? thanks

**Reply by tjboldt** [reply]
I created a new release for D0029-FW000F-HW6 and added the .hdv image to the release itself. It used to be included in the source but I am trying to avoid adding binary drive images to the source itself going forward. This I'm sure contributed to the confusion in the instructions which were written when the image was in the source.

**Comment by k6lcm** [comment]
Where can I find more information about step 2. "Solder chips, header and capacitors in place?" Do you have a link to a BoM?

**Reply by JPT77** [reply]
The parts are written on the PCB.

**Reply by k6lcm** [reply]
That's easy! Thanks.

**Reply by JPT77** [reply]
Looks like you may replace 74LS with 74HCT if you cant source them (works for me at least). 74LVC needs to be LVC because of different voltage levels. C are all 100nF ceramic.

**Reply by microbmen** [reply]
@k6lcm t also updated with a BoM 😄 https://github.com/tjboldt/Apple2-IO-RPi/blob/main/Hardware/BOM.csv

**Reply by tjboldt** [reply]
Thanks @microbmen and @JPT77 for answering questions from people new to the project.

**Comment by CygnusTM** [comment]
I'm working on getting one of these running. I've followed the steps above. When I boot the card, it sees the Pi and seems to start OK, but I get "Apple2-IO-RPi not found" twice then a BASIC prompt. I get the same error when I run "-shell". Any ideas?

**Reply by tjboldt** [reply]
The update from last week requires the latest firmware to match the latest shell, rpi command and clock driver. Unfortunately they are not cross compatible so it’s a full upgrade required this time. Also, only tested with the classic hardware so far.

**Reply by tjboldt** [reply]
Also, probably need to assemble the latest firmware…

cd ~/Apple2-IO-RPi/Apple2
./assemble.sh

**Reply by CygnusTM** [reply]
I've made progress, but not quite there yet. I assembled and reflashed the new firmware, and tried it. It hung up while booting saying "Waiting for RPi..........." So I ran the install script on the Pi again, and now it fully boots, and "-shell" works. 
But now no "RPI" command will work. Whatever command I try to send to the Pi, I get nothing but a stream of garbage that will only stop with a reset.

**Reply by tjboldt** [reply]
Is this the Pico version of the card? I didn’t test the latest changes with it. Likely it will need the offsets fixed.

**Reply by CygnusTM** [reply]
Nope. Regular Pi version using a Pi Zero 2. Board is marked "Sixth Prototype 2022-Feb-20".

**Reply by CygnusTM** [reply]
Another thing that is flaky. The apple2driver.service fails on autostart every time the Pi reboots. Then it will work fine if I restart the service. I might try a fresh build on the Pi next.

```
todd@raspberrypi:~ $ sudo systemctl status apple2driver.service
× apple2driver.service - Apple2-IO-RPi Driver (Classic edition)
     Loaded: loaded (/etc/systemd/system/apple2driver.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Mon 2024-04-29 10:49:28 EDT; 56s ago
   Duration: 777ms
    Process: 392 ExecStart=/home/todd/Apple2-IO-RPi/RaspberryPi/apple2driver/apple2driver (code=exited, status=2)
   Main PID: 392 (code=exited, status=2)
        CPU: 226ms

Apr 29 10:49:28 raspberrypi apple2driver[392]: Starting Apple II RPi v002D...
Apr 29 10:49:28 raspberrypi apple2driver[392]: Starting communications via GPIO ports...
Apr 29 10:49:28 raspberrypi apple2driver[392]: panic: failed to open gpio
Apr 29 10:49:28 raspberrypi apple2driver[392]: goroutine 1 [running]:
Apr 29 10:49:28 raspberrypi apple2driver[392]: github.com/tjboldt/Apple2-IO-RPi/RaspberryPi/apple2driver/a2io.A2Gpio.Init({})
Apr 29 10:49:28 raspberrypi apple2driver[392]:         /home/todd/Apple2-IO-RPi/RaspberryPi/apple2driver/a2io/gpio.go:54 +0x3d0
Apr 29 10:49:28 raspberrypi apple2driver[392]: main.main()
Apr 29 10:49:28 raspberrypi apple2driver[392]:         /home/todd/Apple2-IO-RPi/RaspberryPi/apple2driver/driver.go:60 +0x148
Apr 29 10:49:28 raspberrypi systemd[1]: apple2driver.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Apr 29 10:49:28 raspberrypi systemd[1]: apple2driver.service: Failed with result 'exit-code'.
```

**Reply by CygnusTM** [reply]
Tried a fresh install on the Pi. Both problems persist.

**Reply by tjboldt** [reply]
Strange. The error you’re getting is failing to open the GPIO ports. I would assume that if it’s a permission issue, it would continue to fail to start unless you’re restarting the service as another user with greater privilege after.

**Reply by dkgeorge** [reply]
Was having the same issue and included the line below the [Service] section of /etc/systemd/system/apple2driver.service 

[Service]
ExecStartPre=/bin/sleep 30

This fixed the problem. Apparently apple2driver.service was attempting to start too early during boot.

**Reply by tjboldt** [reply]
I was going to incorporate your workaround but that would make it a lot slower to boot (sometimes a full cold boot is ready in under 30 seconds)... I'll investigate how to check for GPIO ready state or perhaps add a retry on failure

**Reply by tjboldt** [reply]
Moved this to https://github.com/tjboldt/Apple2-IO-RPi/issues/175

**Reply by CygnusTM** [reply]
That workaround solved my service startup issue, but I'm still getting a stream of garbage with every RPI command.

**Reply by CygnusTM** [reply]
Also, none of the a2 commands work. No garbage with those, but nothing happens. The cursor just sits there.

**Reply by tjboldt** [reply]
Are you running the latest firmware? See https://github.com/tjboldt/Apple2-IO-RPi/commit/9786590f6e3a084a0da933964607a4a2162ac93d where something similar was fixed

**Reply by tjboldt** [reply]
For more info, read #172

**Reply by CygnusTM** [reply]
I appeared to be on the latest firmware (0011), but did a full update anyway. Still getting garbage when in slot 5. I've moved it to slot 7 for now, and it is working.

**Reply by tjboldt** [reply]
I’ll have another look at it… maybe the fix wasn’t the full fix.

**Reply by tjboldt** [reply]
https://github.com/tjboldt/Apple2-IO-RPi/assets/21348575/bc1b8191-65ea-4b7a-9502-5c94090dea28

Hmm... seems to work for me but that is a typical developer excuse "works on my machine". Sigh.

**Comment by tjboldt** [comment]
The latest instructions moved to https://github.com/tjboldt/Apple2-IO-RPi/blob/main/ClassicSetup.md so you can open an issue or make a pull request with fixes.

**Comment by kellis5137** [comment]
I was looking at your work, looking at the cost of the board, and noticing that PCBWay offers assembly service. Seems really affordable the point is that it seems too cheap. Is there enough information in the gerber files for them to assemble the board itself? Has anyone tried this? Keith

**Comment by mevenson** [comment]
I just finished the build out of the Apple2-IO-RPi board and installed it in my Apple IIGS. I had to fiddle with where to place it in the slots, but I finally got the follwomg configuraiton to work: Serial card in slot 2, Apple2-IO-RPi in slot 4 and hard drive in slot 6. I can set the CMOS to boot from either the Apple2-IO-RPi in slot 4 or the hard drive in slot 6, I can also set the COMS to scan for boot disk and it boots from slot 4. This is all GREAT, but it would be really nice if I could get the contents of the Hard drive (contains IIGS System 6.0.1) to an image that I could mount on the Apple2-IO-RPi as the boot drive. Any ideas on how I would accomplish this? Also - for some reason, where I boot System 6.0.1 from the hard drive with the Apple2-IO-RPi board installed, the mouse is dead. If I remove the board and reboot - the mouse comes alive. Is that because he mouse and the Apple2-IO-RPi board are both set up for slot 4? I will try yje Apple2-IO-RPi in slot 5 and see if that helps.
